coalib/core/Bear.py

```diff
@@ -0,0 +1,501 @@
+        >>> bear = Bear(section, file_dict)
+
+        We can now carelessly query for a neat file that doesn't exist yet:
+
+        >>> from os import remove
+        >>> if exists(join(bear.data_dir, 'a_file')):
+        ...     remove(join(bear.data_dir, 'a_file'))
+        >>> file = bear.download_cached_file('https://github.com/', 'a_file')
+
+        If we download it again, it'll be much faster as no download occurs:
+
+        >>> newfile = bear.download_cached_file(
+        ...     'https://github.com/', 'a_file')
+        >>> newfile == file
+        True
+
+        :param url:
+            The URL to download the file from.
+        :param filename:
+            The filename it should get, e.g. "test.txt".
+        :return:
+            A full path to the file ready for you to use!
+        """
+        filename = join(cls.data_dir, filename)
+        if exists(filename):
+            return filename
+
+        logging.info('{}: Downloading {} into {!r}.'
+                     .format(cls.name, url, filename))
+
+        response = requests.get(url, stream=True, timeout=20)
+        with open(filename, 'wb') as file:
+            for chunk in response.iter_content(125):
+                file.write(chunk)
+        return filename
+
+    @classproperty
+    def data_dir(cls):
+        """
+        Returns a directory that may be used by the bear to store stuff. Every
+        bear has an own directory dependent on their name.
+        """
+        data_dir = abspath(join(user_data_dir('coala-bears'), cls.name))
+
+        makedirs(data_dir, exist_ok=True)
+        return data_dir
+
+    @property
+    def new_result(self):
+        """
+        Returns a partial for creating a result with this bear already bound.
+        """
+        return partial(Result.from_values, self)
+
+    def execute_task(self, args, kwargs):
+        """
+        Executes a task.
+
+        By default returns a list of results collected from this bear.
+
+        This function has to return something that is picklable to make bears
+        work in multi-process environments.
+
+        :param args:
+            The arguments of a task.
+        :param kwargs:
+            The keyword-arguments of a task.
+        :return:
+            A list of results from the bear.
+        """
+        return list(self.analyze(*args, **kwargs))
+
+    def analyze(self, *args, **kwargs):
+        """
+        Performs the code analysis.
+
+        :return:
+            An iterable of results.
+        """
+        raise NotImplementedError('This function has to be implemented for a '
+                                  'runnable bear.')
+
+    def generate_tasks(self):
+        """
+        This method is responsible for providing the job arguments ``analyze``
+        gets called with.
+
+        :return:
+            An iterable containing the positional and keyword arguments
+            organized in pairs: ``(args-tuple, kwargs-dict)``
```

Comment:
I was thinking of something like creating a new bear and implementing on the method, but it looks like its more suitable for the tests